conjure-runtime/src/service/node/limiter/ciad.rs

```diff
@@ -0,0 +1,297 @@
+use futures::ready;
+use http::{Response, StatusCode};
+use pin_project::pin_project;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const INITIAL_LIMIT: usize = 20;
+const BACKOFF_RATIO: f64 = 0.9;
+const MIN_LIMIT: usize = 1;
+const MAX_LIMIT: usize = 1_000_000;
+
+/// An asynchronous concurrency limiter. "Ciad" stands for "cautious increase, agressive decrease", which is a mouthful!
+///
+/// The limiter is tagged with a "behavior" which controls how a response from the server affects the limit.
+pub struct CiadConcurrencyLimiter<B> {
+    semaphore: Arc<DeficitSemaphore>,
+    limit: AtomicF64,
+    in_flight: AtomicUsize,
+    _p: PhantomData<B>,
+}
+
+impl<B> CiadConcurrencyLimiter<B> {
+    pub fn new() -> Arc<Self> {
+        Arc::new(CiadConcurrencyLimiter {
+            semaphore: DeficitSemaphore::new(INITIAL_LIMIT),
+            limit: AtomicF64::new(INITIAL_LIMIT as f64),
+            in_flight: AtomicUsize::new(0),
+            _p: PhantomData,
+        })
+    }
+
+    pub fn limit(&self) -> f64 {
+        self.limit.load(Ordering::SeqCst)
+    }
+
+    pub fn in_flight(&self) -> usize {
+        self.in_flight.load(Ordering::SeqCst)
+    }
+
+    pub fn acquire(self: Arc<Self>) -> Acquire<B> {
+        Acquire {
+            future: self.semaphore.clone().acquire(),
+            limiter: self,
+        }
+    }
+
+    fn update_limit<F>(&self, f: F)
+    where
+        F: Fn(f64) -> Option<f64>,
+    {
+        let mut old_limit = self.limit.load(Ordering::SeqCst);
+
+        loop {
+            let new_limit = match f(old_limit) {
+                Some(new_limit) => new_limit,
+                None => return,
+            };
+
+            match self.limit.compare_exchange(
+                old_limit,
+                new_limit,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => {
+                    let old_limit = old_limit as usize;
+                    let new_limit = new_limit as usize;
+
+                    #[allow(clippy::comparison_chain)]
+                    if new_limit > old_limit {
+                        self.semaphore.add_permits(new_limit - old_limit);
+                    } else if old_limit > new_limit {
+                        self.semaphore.remove_permits(old_limit - new_limit);
+                    }
+
+                    return;
+                }
+                Err(limit) => old_limit = limit,
+            }
+        }
+    }
+}
+
+pub trait Behavior {
+    fn on_success<B>(response: &Response<B>) -> Mode;
+
+    fn on_failure(error: &Error) -> Mode;
+}
+
+pub enum HostLevel {}
+
+impl Behavior for HostLevel {
+    fn on_success<B>(response: &Response<B>) -> Mode {
+        match response.status() {
+            StatusCode::INTERNAL_SERVER_ERROR | StatusCode::TOO_MANY_REQUESTS => Mode::Ignore,
```

Comment:
Shouldn't `TOO_MANY_REQUESTS` be dropped and `code.is_server_error()` be ignored?

Reply:
Unless conjure switched the meaning of ignored and dropped, these are the meanings defined by the Netflix concurrency-limits library: https://github.com/Netflix/concurrency-limits/blob/master/concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/Limiter.java#L31-L50

Reply:
This is mirroring the behavior in Dialogue: https://github.com/palantir/dialogue/blob/develop/dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java#L85.

The error cases are basically split between the host-level limiter and endpoint-level limiter so things don't get double counted I think.

Reply:
Huh, interesting decision to not count 429s against the host. I guess it's better to be consistent with dialogue.

Reply:
Services (alta) tend to limit individual clients using 429s on individual endpoints, for example read endpoints may not be limited while there are permits available for writes. There's definitely room for improvement in this area :-)

Reply:
I think the theory is that a 429 is a more "intentional" decision by the server and this allows the client to assume it is more granular than a global host-level thing.

It seems like ideally we could have some kind of response header saying what in particular is covered by the 429 though.